File: benchmark/mandelbrot/mandelbrot.js
// Mandelbrot set benchmark for JavaScript
// Counts total iterations across a grid of points
// Real-world use case: fractal rendering, GPU benchmarks
//
// Reports throughput (pixels rendered per second). The iteration count
// auto-calibrates so the timed loop runs for about a second.
//
// How to run:
//   mise run benchmark-mandelbrot
//
// Or manually:
//   node benchmark/mandelbrot/mandelbrot.js

const TARGET_NS = 1_000_000_000; // ~1s budget

function nowNs() {
  return Math.round(performance.now() * 1e6);
}

function nextIters(n, elapsed, target) {
  const e = elapsed > 0 ? elapsed : 1;
  let est = Math.floor((n * target) / e);
  const hi = n * 100;
  if (est > hi) est = hi;
  if (est > 1_000_000_000) est = 1_000_000_000;
  if (est < 1) est = 1;
  return est;
}

function printThroughput(workPerIter, n, elapsedNs, unit) {
  const secs = elapsedNs / 1e9;
  const rate = secs > 0 ? (workPerIter * n) / secs : 0;
  const perMs = elapsedNs / n / 1e6;
  let rbuf;
  if (unit === "B") {
    if (rate >= 1e9) rbuf = `${(rate / 1e9).toFixed(2)} GB/s`;
    else if (rate >= 1e6) rbuf = `${(rate / 1e6).toFixed(2)} MB/s`;
    else if (rate >= 1e3) rbuf = `${(rate / 1e3).toFixed(2)} KB/s`;
    else rbuf = `${rate.toFixed(2)} B/s`;
  } else {
    if (rate >= 1e9) rbuf = `${(rate / 1e9).toFixed(2)} G ${unit}/s`;
    else if (rate >= 1e6) rbuf = `${(rate / 1e6).toFixed(2)} M ${unit}/s`;
    else if (rate >= 1e3) rbuf = `${(rate / 1e3).toFixed(2)} k ${unit}/s`;
    else rbuf = `${rate.toFixed(2)} ${unit}/s`;
  }
  console.log(`Throughput: ${rbuf}   (${perMs.toFixed(3)} ms/iter, ${n} iter)`);
}

function mandelbrotIterations(cx, cy, maxIter) {
  let x = 0.0;
  let y = 0.0;
  let iter = 0;

  while (iter < maxIter) {
    const x2 = x * x;
    const y2 = y * y;

    if (x2 + y2 > 4.0) {
      return iter;
    }

    const xy = x * y;
    x = x2 - y2 + cx;
    y = 2.0 * xy + cy;
    iter += 1;
  }

  return maxIter;
}

function mandelbrotTotal(width, height, maxIter) {
  // Mandelbrot region: x in [-2.5, 1.0], y in [-1.0, 1.0]
  const xMin = -2.5;
  const xMax = 1.0;
  const yMin = -1.0;
  const yMax = 1.0;

  const dx = (xMax - xMin) / width;
  const dy = (yMax - yMin) / height;

  let totalIterations = 0;

  for (let py = 0; py < height; py++) {
    const cy = yMin + py * dy;

    for (let px = 0; px < width; px++) {
      const cx = xMin + px * dx;
      totalIterations += mandelbrotIterations(cx, cy, maxIter);
    }
  }

  return totalIterations;
}

function main() {
  const width = 1024;
  const height = 768;
  const maxIter = 256;

  // Warmup.
  let total = mandelbrotTotal(width, height, maxIter);

  let n = 1;
  let elapsed = 0;
  for (;;) {
    const start = nowNs();
    for (let i = 0; i < n; i++) {
      total = mandelbrotTotal(width, height, maxIter);
    }
    elapsed = nowNs() - start;
    if (elapsed >= TARGET_NS) break;
    const nx = nextIters(n, elapsed, TARGET_NS);
    if (nx <= n) break;
    n = nx;
  }

  console.log(`Mandelbrot ${width}x${height}, max_iter=${maxIter}`);
  console.log(`Total iterations: ${total}`);
  printThroughput(width * height, n, elapsed, "px");
}

main();

File: benchmark/mandelbrot/mandelbrot.c
// Mandelbrot set benchmark for C
// Counts total iterations across a grid of points
// Real-world use case: fractal rendering, GPU benchmarks
//
// Reports throughput (pixels rendered per second). The iteration count
// auto-calibrates so the timed loop runs for about a second.
//
// How to run:
//   mise run benchmark-mandelbrot
//
// Or manually:
//   cc -O3 -ffp-contract=off -o benchmark/mandelbrot/mandelbrot_c benchmark/mandelbrot/mandelbrot.c
//   ./benchmark/mandelbrot/mandelbrot_c
//
// Note: -ffp-contract=off disables FMA (fused multiply-add) to ensure
// consistent IEEE 754 behavior with JavaScript and WebAssembly.

#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <time.h>

#define TARGET_NS 1000000000LL  // ~1s budget

static long long now_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return (long long)t.tv_sec * 1000000000LL + t.tv_nsec;
}

static long long next_iters(long long n, long long elapsed, long long target) {
    long long e = elapsed > 0 ? elapsed : 1;
    long long est = n * target / e;
    long long hi = n * 100;
    if (est > hi) est = hi;
    if (est > 1000000000LL) est = 1000000000LL;
    if (est < 1) est = 1;
    return est;
}

static void print_throughput(double work_per_iter, long long n, long long elapsed_ns, const char *unit) {
    double secs = (double)elapsed_ns / 1e9;
    double rate = secs > 0.0 ? (work_per_iter * (double)n) / secs : 0.0;
    double per_ms = (double)elapsed_ns / (double)n / 1e6;
    char rbuf[64];
    if (strcmp(unit, "B") == 0) {
        if (rate >= 1e9) snprintf(rbuf, sizeof rbuf, "%.2f GB/s", rate / 1e9);
        else if (rate >= 1e6) snprintf(rbuf, sizeof rbuf, "%.2f MB/s", rate / 1e6);
        else if (rate >= 1e3) snprintf(rbuf, sizeof rbuf, "%.2f KB/s", rate / 1e3);
        else snprintf(rbuf, sizeof rbuf, "%.2f B/s", rate);
    } else {
        if (rate >= 1e9) snprintf(rbuf, sizeof rbuf, "%.2f G %s/s", rate / 1e9, unit);
        else if (rate >= 1e6) snprintf(rbuf, sizeof rbuf, "%.2f M %s/s", rate / 1e6, unit);
        else if (rate >= 1e3) snprintf(rbuf, sizeof rbuf, "%.2f k %s/s", rate / 1e3, unit);
        else snprintf(rbuf, sizeof rbuf, "%.2f %s/s", rate, unit);
    }
    printf("Throughput: %s   (%.3f ms/iter, %lld iter)\n", rbuf, per_ms, n);
}

int mandelbrot_iterations(double cx, double cy, int max_iter) {
    double x = 0.0;
    double y = 0.0;
    int iter = 0;

    while (iter < max_iter) {
        double x2 = x * x;
        double y2 = y * y;

        if (x2 + y2 > 4.0) {
            return iter;
        }

        double xy = x * y;
        x = x2 - y2 + cx;
        y = 2.0 * xy + cy;
        iter += 1;
    }

    return max_iter;
}

int64_t mandelbrot_total(int width, int height, int max_iter) {
    // Mandelbrot region: x in [-2.5, 1.0], y in [-1.0, 1.0]
    double x_min = -2.5;
    double x_max = 1.0;
    double y_min = -1.0;
    double y_max = 1.0;

    double dx = (x_max - x_min) / (double)width;
    double dy = (y_max - y_min) / (double)height;

    int64_t total_iterations = 0;

    for (int py = 0; py < height; py++) {
        double cy = y_min + (double)py * dy;

        for (int px = 0; px < width; px++) {
            double cx = x_min + (double)px * dx;
            total_iterations += mandelbrot_iterations(cx, cy, max_iter);
        }
    }

    return total_iterations;
}

int main() {
    volatile int width = 1024;
    volatile int height = 768;
    volatile int max_iter = 256;

    // Warmup.
    int64_t total = mandelbrot_total(width, height, max_iter);

    long long n = 1, elapsed = 0;
    for (;;) {
        long long start = now_ns();
        for (long long i = 0; i < n; i++) {
            total = mandelbrot_total(width, height, max_iter);
        }
        elapsed = now_ns() - start;
        if (elapsed >= TARGET_NS) break;
        long long nx = next_iters(n, elapsed, TARGET_NS);
        if (nx <= n) break;
        n = nx;
    }

    printf("Mandelbrot %dx%d, max_iter=%d\n", width, height, max_iter);
    printf("Total iterations: %lld\n", (long long)total);
    print_throughput((double)width * (double)height, n, elapsed, "px");

    return 0;
}
